bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/client/RpcClient.java

```diff
@@ -83,14 +85,35 @@ public RpcClient(HomematicConfig config) {
     /**
      * Register a callback for the specified interface where the Homematic gateway can send its events.
      */
-    public void init(HmInterface hmInterface, String clientId) throws IOException {
+    public void init(HmInterface hmInterface) throws IOException {
         RpcRequest<T> request = createRpcRequest("init");
         request.addArg(getRpcCallbackUrl());
-        request.addArg(clientId);
+        request.addArg(thisUID);
         if (config.getGatewayInfo().isHomegear()) {
             request.addArg(Integer.valueOf(0x22));
         }
-        sendMessage(config.getRpcPort(hmInterface), request);
+        int maxRetryAttempts = config.getCallbackRegistrationRetries();
+        int waitTime = config.getRetryWaitTime() * 1000;
+        for (int attempt = 1; attempt <= maxRetryAttempts; attempt++) {
+            try {
+                logger.debug("Register callback for interface {}, attempt {}/{}", hmInterface.getName(), attempt,
+                        maxRetryAttempts);
+                sendMessage(config.getRpcPort(hmInterface), request);
+                break;
+            } catch (IOException e) {
+                if (attempt == maxRetryAttempts) {
+                    logger.warn(
+                            "All attempts to register a callback for interface '{}' failed, last exception message: {}",
+                            hmInterface.getName(), e.getMessage());
+                    throw (e);
+                }
+                try {
+                    Thread.sleep(waitTime);
```

Comment:
The ThingHandler's `initialize()` method must return fast. As this code is called via that method, it must also return fast. Despite that, sleep should be avoided, because it blocks a thread in the schedulers thread pool. You could schedule another try via `scheduler.scheduleWithFixedDelay()`, if `sendMessage()` fails.

Reply:
Thanks for the review and the tips.
The "fast return" is guaranteed in my opinion, because this method is called via `scheduler.submit`: 
https://github.com/openhab/openhab-addons/blob/e24d43aecbcfc5af551c8e9087d9be6992e5c69a/bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicBridgeHandler.java#L95

I will replace the `sleep`.

Reply:
@fwolter I have changed the implementation and replaced the sleep and implemented it as you suggested.